ADR/0033-upstreaming-integration-service.md

```diff
@@ -0,0 +1,228 @@
+
+## Decision
+
+- **App Deployment** will be dropped, per the [Decoupling Deployment] ADR.
+- **Release Initiation** will become the responsibility of [release-service]. See more below.
+- [integration-service] will continue to own **Test Initiation**, **DeploymentTarget Provisioning**,
+  **Snapshot Test Results Recording**, and **Source Repo Test Results Recording**.
+- **Snapshot Construction** and **Global Candidate List Maintenance** will remain responsibilities
+  of [integration-service], but will undergo some changes to remove dependence on external APIs.
+
+### The Snapshot becomes the TestSubject
+
+[integration-service] will fork the [Snapshot] API to a new resource under its domain
+named [TestSubject].
+
+The purpose of this change is to make [integration-service] independent of other APIs. The name
+[TestSubject] is chosen to reflect what the [Snapshot] represented *in [integration-service]'s
+domain* in our original architecture.
+
+### References to the Application become a label or selector
+
+Today, you can tell which [Application] a [IntegrationTestScenario] and [Snapshot] refer to by way
+of an `application` reference in the `spec`. However, with our goal of decoupling
+[integration-service] from the rest of the controllers in mind, we need to remove this reference.
+
+Instead of adding an `application` reference to a [TestSubject], use a label called `LABELNAME` to
+identify which [IntegrationTestScenarios] it is relevant to.
+
+On the [IntegrationTestScenario], drop the `application` from the spec and replace it with
+a [selector] which the user can use to reference the `LABELNAME` label on [TestSubjects]. It should
+be possible for the user to specify a selector that references any attribute they wish. The user
+should conclude naturally that `LABELNAME` is a good candidate for usage in their selectors. UI and
+docs can guide this.
+
+- A new [TestSubject] without a `LABELNAME` label should be mutated by webhook to gain a value of
+  "default".
+- Attempts to remove the `LABELNAME` label from a [TestSubject] should be rejected (or mutated to
+  "default"?)
+- Attempts to edit the `LABELNAME` label on a [TestSubject] should be permitted.
+
+### The Global Candidate List becomes the control TestSubject, the last TestSubject to pass testing
+
+Today, the "global candidate list" is a term that we use to refer to the list of all the built-image
+pullspecs attached to every [Component] of an [Application]. Replace that implementation
+with an identifying *label* on a [TestSubject] in the namespace: `CONTROLLABEL`. Call the [TestSubject] that
+carries this `CONTROLLABEL` label the **control** TestSubject for the `LABELNAME` label.
+
+[integration-service] should grow a mutating webhook that *removes* the `CONTROLLABEL` label from
+the last control [TestSubject] for the `LABELNAME` label, when the `CONTROLLABEL` label is applied
+to a new [TestSubject] bearing the `LABELNAME` label.
```

Comment:
Would it be acceptable to re-apply the `CONTROLLABEL` on a TestSubject that previously had the label but has been removed? Would there need to be any additional validation on such a TestSubject?

Reply:
Yes, I think it would be acceptable - to reset your reference subject back to whatever it was at some point in the past. No additional validation necessary from my PoV.